### PR TITLE
Feat/set initial case answers

### PR DESCRIPTION
**Pull request by jonatanhanson**
## Explain the changes you’ve made
Added functionality to pre populate case answers with previous case answers and user information. 

## Explain your solution
Created a method that takes four arguments (initial forms, user object, previous case & form templates) and populates the initial forms with previous case data and user data. 
Which fields that will be pre populated can be defined from the formbuilder.

**Note** that the feature to load previous data is only activated to some of the fields(text, number, checkbox, repeaterField and editableList) in the formbuilder and will not work for rest of the fields. I'm not sure why we don't have it activated for all fields, could be a future task to fix. 

This new functionality is used by createCase + createVivaCase lambdas during the creation of new cases.

I have chosen to get previous case data from the latest updated case that also has status set to "closed". So this will not get previous answers from "ongoing" cases for example.

## How to test the changes?

First make sure you have a user, forms and a case(with status: closed) in your DB.
The form needs to have fields with "load previous" functionality activated. 

Deploy case and viva services. 
Make a POST request to create case endpoint.
Example body json:
`
{
  "statusType": "notStarted",
  "currentFormId": "SOME-FORM-ID",
  "provider": "VIVA",
  "details": {},
  "forms": {
    "SOME-FORM-ID {
      "answers": [],
      "currentPosition": {
        "index": 1,
        "level": 0,
        "currentMainStep": 0,
        "currentMainStepIndex": 0
      }
    }
  }
}
` 

If everything works the answers object should be populated with the data you have defined in form builder.  

## Anything else
I have only tested this with the createCase endpoint, because I dont know how to trigger event that creates a new VIVA case. So it would be great if someone could test that too.

### CODE REVIEW

**Comment by zilaei** [comment]
Tested with the app and can verify that previews data is loaded in current case (for dynamic inputs).

File: services/viva-ms/lambdas/createVivaCase.js [hunk 1]
```diff
@@ -152,3 +158,70 @@ async function putRecurringVivaCase(PK, workflowId, period) {
...
+    Key: {
+      personalNumber,
+    },
+  };
+
+  const [error, dbResponse] = await to(dynamoDB.call('get', params));
+  if (!dbResponse) {
+    console.error('(cases-api) DynamoDb query on users table failed', error);
+    return;
+  }
```
Comment (TeddyAndersson) [inline services/viva-ms/lambdas/createVivaCase.js:175]: I would rather see that we throw this error than return nothing at all. the console.error can be used in an catch further upstream.

File: services/viva-ms/lambdas/createVivaCase.js [hunk 2]
```diff
@@ -152,3 +158,70 @@ async function putRecurringVivaCase(PK, workflowId, period) {
...
+      ':statusTypeClosed': 'closed',
+      ':provider': provider,
+    },
+  };
+
+  const [error, dbResponse] = await to(dynamoDB.call('query', params));
+  if (!dbResponse) {
+    console.error('(cases-api) DynamoDb query on cases table failed', error);
+    return;
+  }
```
Comment (TeddyAndersson) [inline services/viva-ms/lambdas/createVivaCase.js:222]: Throw error instead of return.

File: services/viva-ms/lambdas/createVivaCase.js [hunk 3]
```diff
@@ -126,6 +127,11 @@ async function putRecurringVivaCase(PK, workflowId, period) {
     [completionFormId]: initialFormAttributes,
   };
 
+  const user = await getUser(PK);
+  const formTemplates = await getFormTemplates(initialForms);
+  const previousCase = await getLastUpdatedCase(PK, CASE_PROVIDER_VIVA);
```
Comment (TeddyAndersson) [inline services/viva-ms/lambdas/createVivaCase.js:132]: use await to syntax

Comment (TeddyAndersson) [reply services/viva-ms/lambdas/createVivaCase.js:132]: and logg the errors here in the main function

**Review by TeddyAndersson — CHANGES_REQUESTED** [review]
The code works from the test that i made on my personal environment. Added some comments for code readability and consistency

File: services/cases-api/lambdas/createCase.js [hunk 4]
```diff
@@ -104,3 +124,67 @@ async function parseJsonD(data) {
...
+    TableName: config.users.tableName,
+    Key: {
+      personalNumber,
+    },
+  };
+
+  const [error, dbResponse] = await to(dynamoDb.call('get', params));
+  if (!dbResponse) {
+    throwError(error.statusCode, error.message);
+  }
```
Comment (dannil76) [inline services/cases-api/lambdas/createCase.js:139]: To be consistent with how other lambdas is handling error we can write this like
`if (error) {
.....
}`

:)

File: services/cases-api/lambdas/createCase.js [hunk 5]
```diff
@@ -104,3 +124,67 @@ async function parseJsonD(data) {
...
+      Key: {
+        PK: `FORM#${key}`,
+      },
+    };
+    const [error, dbResponse] = await to(dynamoDb.call('get', params));
+
+    if (!dbResponse) {
+      console.error('(cases-api) DynamoDb query on forms table failed', error);
+      continue;
+    }
```
Comment (dannil76) [inline services/cases-api/lambdas/createCase.js:158]: same comment as previous comment :)

File: services/cases-api/lambdas/createCase.js [hunk 6]
```diff
@@ -104,3 +124,67 @@ async function parseJsonD(data) {
...
+      ':pk': PK,
+      ':statusTypeClosed': 'closed',
+      ':provider': provider,
+    },
+  };
+
+  const [error, dbResponse] = await to(dynamoDb.call('query', params));
+  if (!dbResponse) {
+    throwError(error.statusCode, error.message);
+  }
```
Comment (dannil76) [inline services/cases-api/lambdas/createCase.js:185]: same as previous comments :)

File: services/viva-ms/lambdas/createVivaCase.js [hunk 7]
```diff
@@ -152,3 +172,68 @@ async function putRecurringVivaCase(PK, workflowId, period) {
...
+    TableName: config.users.tableName,
+    Key: {
+      personalNumber,
+    },
+  };
+
+  const [error, dbResponse] = await to(dynamoDB.call('get', params));
+  if (!dbResponse) {
+    throwError(error.statusCode, error.message);
+  }
```
Comment (dannil76) [inline services/viva-ms/lambdas/createVivaCase.js:188]: same feedback as previous comments on how error handling is done in other lambdas
check IF error is a little more telling :)

File: services/viva-ms/lambdas/createVivaCase.js [hunk 8]
```diff
@@ -152,3 +172,68 @@ async function putRecurringVivaCase(PK, workflowId, period) {
...
+      Key: {
+        PK: `FORM#${key}`,
+      },
+    };
+    const [error, dbResponse] = await to(dynamoDB.call('get', params));
+
+    if (!dbResponse) {
+      console.error('(cases-api) DynamoDb query on forms table failed', error);
+      continue;
+    }
```
Comment (dannil76) [inline services/viva-ms/lambdas/createVivaCase.js:207]: same feedback as previous

File: services/viva-ms/lambdas/createVivaCase.js [hunk 9]
```diff
@@ -152,3 +172,68 @@ async function putRecurringVivaCase(PK, workflowId, period) {
...
+      ':pk': PK,
+      ':statusTypeClosed': 'closed',
+      ':provider': provider,
+    },
+  };
+
+  const [error, dbResponse] = await to(dynamoDB.call('query', params));
+  if (!dbResponse) {
+    throwError(error.statusCode, error.message);
+  }
```
Comment (dannil76) [inline services/viva-ms/lambdas/createVivaCase.js:234]: same feedback as previous

**Review by dannil76 — CHANGES_REQUESTED** [review]
some minor aesthetically request :)

File: services/cases-api/lambdas/createCase.js (hunk 4 above)
Comment (jonatanhanson) [reply services/cases-api/lambdas/createCase.js]: Makes sense, fixed now

**Review by dannil76 — APPROVED** [review]
Brilliant :)

**Pull request merged**